cpu/stm32/periph/timer.c

```diff
@@ -146,6 +147,46 @@ int timer_set_absolute(tim_t tim, int channel, unsigned int value)
     return 0;
 }
 
+int timer_set(tim_t tim, int channel, unsigned int timeout)
+{
+    if (channel >= (int)TIMER_CHANNEL_NUMOF) {
+        return -1;
+    }
+
+    unsigned irqstate = irq_disable();
+    set_oneshot(tim, channel);
+
+    /* clear spurious IRQs */
+    dev(tim)->SR &= ~(TIM_SR_CC1IF << channel);
+
+    unsigned value = (dev(tim)->CNT + timeout) & timer_config[tim].max;
+    TIM_CHAN(tim, channel) = value;
+
+    /* enable IRQ */
+    dev(tim)->DIER |= (TIM_DIER_CC1IE << channel);
+
+#ifdef MODULE_PERIPH_TIMER_PERIODIC
+    if (dev(tim)->ARR == TIM_CHAN(tim, channel)) {
+        dev(tim)->ARR = timer_config[tim].max;
+    }
+#endif
+
+    /* calculate time till timeout */
+    value = (value - dev(tim)->CNT) & timer_config[tim].max;
+
+    if (value > timeout) {
+        /* time till timeout is larger than requested --> timer already expired
+         * ==> let's make sure we have an IRQ pending :) */
+        while (!(dev(tim)->SR & (TIM_SR_CC1IF << channel))) {
+            TIM_CHAN(tim, channel) = dev(tim)->CNT;
+        }
```

Comment:
Are you sure this can't deadlock?

Reply:
When the timer clock gets close to the CPU clock it likely will. But then again, this driver doesn't work anyway with higher frequency (we only test it with what xtimer and ztimer uses...)

I guess that the nRF5x approach is the most reliable, but is messes with time keeping on too short timeouts.

Reply:
Wouldn't `TIM_CHAN(tim, channel) = dev(tim)->CNT + 1;` be more robust?

Reply:
Yes, slightly. The assigment is one load and one store instruction. With the `+1` it would be three cycles to update the timer. So instead of not generating an interrupt if one timer tick happens within two CPU cycles, it would fail at two timer ticks happing within three CPU cycles.

Reply:
I changed it to the approach used in the nRF5x implementation. This should now be rock-solid.